File: EUThreadManager_Objective-C/EUThreadManager.m
//
//  EUThreadManager.m
//  EUThreadManager
//
//  Created by Vitaliy Yarkun on 4/22/17.
//  Copyright © 2017 Vitaliy Yarkun. All rights reserved.
//

#import "EUThreadManager.h"

@implementation EUThreadManager

+(instancetype) sharedInstance {
    
    static EUThreadManager* instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[EUThreadManager alloc] init];
    });
    
    return instance;
}

-(void) runOnBackgroundSimpleTask:(simpleBlock) task withPriority:(TaskPriorityExecution) priority andResultMethod:(SEL) resultMethod fromObject:(id) object{
    
    switch (priority) {
        case kHighPriorityExecution:
            [self runHighPriorityTask:task withResultMethod:resultMethod fromObject:object];
            break;
        case kDefaultPriorityExecution:
            [self runDefaultPriorityTask:task withResultMethod:resultMethod fromObject:object];
            break;
        case kLowPriorityExecution:
            [self runLowPriorityTask:task withResultMethod:resultMethod fromObject:object];
            break;
        case kBackgroundPriorityExecution:
            [self runBackgroundPriorityTask:task withResultMethod:resultMethod fromObject:object];
            break;
        default:
            break;
    }
}

-(void) runHighPriorityTask:(simpleBlock) task withResultMethod:(SEL) resultMethod fromObject:(id) object {
   
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH, 0), ^{
        task();
        dispatch_async(dispatch_get_main_queue(), ^{
            [object performSelector:resultMethod];
        });
    });
}

-(void) runDefaultPriorityTask:(simpleBlock) task withResultMethod:(SEL) resultMethod fromObject:(id) object {
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        task();
        dispatch_async(dispatch_get_main_queue(), ^{
            [object performSelector:resultMethod];
        });
    });
}

-(void) runLowPriorityTask:(simpleBlock) task withResultMethod:(SEL) resultMethod fromObject:(id) object {
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_LOW, 0), ^{
        task();
        dispatch_async(dispatch_get_main_queue(), ^{
            [object performSelector:resultMethod];
        });
    });
}

-(void) runBackgroundPriorityTask:(simpleBlock) task withResultMethod:(SEL) resultMethod fromObject:(id) object {
    
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
        task();
        dispatch_async(dispatch_get_main_queue(), ^{
            [object performSelector:resultMethod];
        });
    });
}

@end

File: EUThreadManager_Swift/EUThreadManager.swift
//
//  EUThreadManager.swift
//  EUThreadManagerSwift
//
//  Created by Vitaliy Yarkun on 4/23/17.
//  Copyright © 2017 Vitaliy Yarkun. All rights reserved.
//

import Foundation

enum TaskPriorityExecution {
    case kUserInitiatedPriorityExecution
    case kDefaultPriorityExecution
    case kUtilityPriorityExecution
    case kBackgroundPriorityExecution
}

typealias simpleTask = ()->()

class EUThreadManager {
    
    private init() {
        
    }
    
    static let sharedInstance = EUThreadManager()
    
    //MARK: Public methods
    func run(task: @escaping simpleTask, with priority: TaskPriorityExecution, and resultMethod: @escaping simpleTask) {
        
        switch priority {
        case .kUserInitiatedPriorityExecution:
            runUserInitiatedPriority(task, with: resultMethod)
            break
        case .kDefaultPriorityExecution:
            runDefaultPriority(task, with: resultMethod)
            break
        case .kUtilityPriorityExecution:
            runUtilityPriority(task, with: resultMethod)
            break
        default:
            runBackgroundPriority(task, with: resultMethod)
            break
        }
    }
    
    private func runUserInitiatedPriority(_ task: @escaping simpleTask, with resultMethod: @escaping simpleTask) {
        
        DispatchQueue.global(qos: .userInitiated).async {
            task()
            DispatchQueue.main.async {
                resultMethod()
            }
        }
    }
    
    private func runDefaultPriority(_ task: @escaping simpleTask, with resultMethod: @escaping simpleTask) {
        
        DispatchQueue.global(qos: .default).async {
            task()
            DispatchQueue.main.async {
                resultMethod()
            }
        }
    }
    
    private func runUtilityPriority(_ task: @escaping simpleTask, with resultMethod: @escaping simpleTask) {
        
        DispatchQueue.global(qos: .utility).async {
            task()
            DispatchQueue.main.async {
                resultMethod()
            }
        }
    }
    
    private func runBackgroundPriority(_ task: @escaping simpleTask, with resultMethod: @escaping simpleTask) {
        
        DispatchQueue.global(qos: .background).async {
            task()
            DispatchQueue.main.async {
                resultMethod()
            }
        }
    }
}
